File: javascript/117.populating-next-right-pointers-in-each-node-ii.js
/*
 * @lc app=leetcode id=117 lang=javascript
 *
 * [117] Populating Next Right Pointers in Each Node II
 */

// @lc code=start
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function(root) {
  // BFS keeping track of order
  // O(n) time
  // O(n) space
  if (!root) return root
  let traversalOrder = []
  let queue = []
  queue.unshift([root, 0])
  while (queue.length > 0) {
    let vals = queue.shift()
    let currentNode = vals[0]
    let level = vals[1]

    if (traversalOrder[level]) {
      traversalOrder[level].push(currentNode)
    } else {
      traversalOrder[level] = [currentNode]
    }

    if (currentNode.left) {
      queue.unshift([currentNode.left, level+1])
    }
    if (currentNode.right) {
      queue.unshift([currentNode.right, level+1])
    }
  }

  for (let i = 0; i < traversalOrder.length; i++) {
    for (let j = traversalOrder[i].length-1; j >= 0; j--) {
      traversalOrder[i][j].next = traversalOrder[i][j-1]
    }
  }

  return root
};
// @lc code=end

File: cpp/117.populating-next-right-pointers-in-each-node-ii.cpp
/*
 * @lc app=leetcode id=117 lang=cpp
 *
 * [117] Populating Next Right Pointers in Each Node II
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution
{
public:
    struct QueueVals
    {
        Node *root;
        int level;
    };
    Node *connect(Node *root)
    {
        // BFS traversal keeping track of all nodes at every level
        // Iterate over nodes, build linked lists
        // O(n) time O(n) space
        if (!root)
            return root;
        queue<QueueVals> myQueue;
        myQueue.push(QueueVals{root, 0});
        vector<vector<Node *>> levelNodes;
        while (myQueue.size() > 0)
        {
            QueueVals vals = myQueue.front();
            myQueue.pop();
            Node *currNode = vals.root;
            int currLevel = vals.level;
            if (currLevel == levelNodes.size())
            {
                // create vector at index
                vector<Node *> newVect{currNode};
                levelNodes.push_back(vector<Node *>{currNode});
            }
            else
            {
                // push to vector at index
                levelNodes[currLevel].push_back(currNode);
            }

            if (currNode->left)
            {
                myQueue.push({currNode->left, currLevel + 1});
            }
            if (currNode->right)
            {
                myQueue.push({currNode->right, currLevel + 1});
            }
        }

        for (vector<Node *> nodes : levelNodes)
        {
            for (int i = 0; i < nodes.size() - 1; i++)
            {
                nodes[i]->next = nodes[i + 1];
            }
        }

        return root;
    }
};
// @lc code=end
